doc/dev/architecture.md

```diff
@@ -0,0 +1,223 @@
+    For more details consult the [next section](#compiler-core).
+
+- **Code Generation:** At this stage the LLVMIR is compiled down to native object code using the
+    LLVM Static Compiler (`llc`) for the local system architecture. A native linker is then used
+    to link the user program to the Catalyst Runtime library.
+    The frontend will load this library into the Python environment and attach its entry point to
+    the callable `@qjit` object defined by the user.
+    For more details on the program execution consult the [runtime section](#runtime--execution).
+
+Elaborating on the program capture phase, tracing is a mechanism by which a function is executed
+with abstract arguments called *tracers*. Calling operations from the tracing library (such as
+`jax.numpy`) will record the operation and which tracers (or constants) it acts upon in a global
+data structure. Note that JAX's tracing contexts are thread-local and can be nested to allow scoped
+region capture, which is relevant when tracing control flow operations.
+
+During the tracing of quantum functions (`qml.qnode`), PennyLane's queuing context is activated to
+build a `QuantumTape` data structure that records all quantum operations. Here as well nested
+queuing contexts are leveraged to allow for scoped operation capture, including control flow
+operations which are themselves captured as pseudo-quantum operations on the tape.
+
+Catalyst provides the "glue" to embed quantum tapes into the JAXPR, by converting PennyLane
+operations to their corresponding JAX primitive and by connecting operation arguments/results to the correct tracer objects.
+
+
+## Compiler Core
+
+This section will focus on the series of compiler passes that convert a high-level quantum program
+to its low-level LLVMIR and subsequent binary form. While the exact passes may frequently change,
+the general stages should still be applicable.
+See the graph below for an overview of the transformations applied to the user program:
+
+![img](../_static/arch/compiler.svg)
+
+- **HLO lowering:**
+
+  - HLO is a high-level IR used by the XLA compiler to compile tensor compute graphs, like those
+    produced by JAX and TensorFlow. JAX natively outputs HLO in MLIR form, via the `StableHLO` MLIR
+    dialect.
+
+  - Since we provide our own compilation & code generation pipeline, we lower out of the HLO dialect
+    into standard MLIR dialects, such as `linalg`, `arith`, `func`, and others. This is done using
+    the lowerings provided by the [mlir-hlo](https://github.com/tensorflow/mlir-hlo) project over
+    in the TensorFlow repository.
+
+  - Quantum dialect operations present in the input are not affected by this transformation.
+
+- **Quantum optimizations:**
+
+  - Many quantum compilation routines can be run at this point, in order to reduce the gate or qubit
+    count of the program. This could include peephole optimzations expressed as MLIR DAG rewrites
```

Comment:
Hmm good idea, do you have a resource or paper in mind? If not I can try and find something.

Reply:
I think even adding a link to Wikipedia is fine here. More so to allow somebody to grab a flavour of the topic, rather than in-depth paper. Though, if an easy to parse paper exists that's also a solid option.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/160/commits/53120fad307eba418d6b280771478a3954c601a1